src/connection/stream.rs

```diff
@@ -175,16 +172,29 @@ impl Stream {
         }
     }
 
-    /// Try to deliver a pending window update.
-    fn send_pending_window_update(&mut self, cx: &mut Context) -> Poll<io::Result<()>> {
-        if self.window_update.is_some() {
+    /// Send new credit to the sending side via a window update message if
+    /// permitted.
+    fn send_window_update(&mut self, cx: &mut Context) -> Poll<io::Result<()>> {
+        // When using [`WindowUpdateMode::OnReceive`] window update messages are
+        // send early on data receival (see [`crate::Connection::on_frame`]).
+        if matches!(self.config.window_update_mode, WindowUpdateMode::OnReceive) {
+            return Poll::Ready(Ok(()));
+        }
+
+        let mut shared = self.shared.lock();
+
+        if let Some(credit) = shared.next_window_update() {
             ready!(self.sender.poll_ready(cx).map_err(|_| self.write_zero_err())?);
```

Comment:
If I understand correctly, then now `Stream::poll_read` and `Stream::poll_next` will effectively block reading on the sending of a window update if half of the receive window has been consumed. That seems unfortunate. How about returning `Poll::Pending` here _only if_ the `shared.window` is `0`? So if `sender.poll_ready()` gives `Poll::Pending` but the window is `> 0` we can continue, as before. Alternatively we may not need to "block" at all on `ready!(send_window_update())` within `Stream::poll_read()` and `Stream::poll_next()`.